mlflow/tensorflow.py

```diff
@@ -333,3 +344,210 @@ def predict(self, df):
+                else tensorflow.keras.backend.eval(opt._epsilon)
+            mlflow.log_param('epsilon', epsilon)
+        l = []
+        self.model.summary(print_fn=l.append)
+        summary = '\n'.join(l)
+        mlflow.set_tag('summary', summary)
+        # TODO: Fix for keras log_model not saving TF optimizers
+        mlflow.keras.log_model(self.model, artifact_path='model')
+
+
+def _log_artifacts_with_warning(**kwargs):
+    try:
+        mlflow.log_artifacts(**kwargs)
+    except MlflowException as e:
+        warnings.warn("Logging to MLflow failed: " + str(e))
+
+
+def _assoc_list_to_map(lst):
+    """
+    Convert an association list to a dictionary.
+    """
+    d = {}
+    for run_id, metric in lst:
+        d[run_id] = d[run_id] + [metric] if run_id in d else [metric]
+    return d
+
+
+def flush_queue():
+    """
+    Flush the metric queue and log contents in batches to MLflow.
+    Queue is divided into batches according to run id.
+    """
+    global _metric_queue
+    try:
+        client = mlflow.tracking.MlflowClient()
+        dic = _assoc_list_to_map(_metric_queue)
+        for key in dic:
+            client.log_batch(key, metrics=dic[key], params=[], tags=[])
+    except MlflowException as e:
+        warnings.warn("Logging to MLflow failed: " + str(e))
+    finally:
+        _metric_queue = []
+
+
+atexit.register(flush_queue)
+
+
+def _add_to_queue(key, value, step, run_id):
+    """
+    Add a metric to the metric queue. Flush the queue if it exceeds
+    max size.
+    """
+    met = Metric(key=key, value=value, timestamp=int(time.time()*1000), step=step)
+    _metric_queue.append((run_id, met))
+    if len(_metric_queue) > _MAX_METRIC_QUEUE_SIZE:
+        flush_queue()
+
+
+def _log_event(event):
+    """
+    Extracts metric information from the event protobuf
+    """
+    if mlflow.active_run() is None:
+        mlflow.start_run()
+    if event.WhichOneof('what') == 'summary':
+        summary = event.summary
+        for v in summary.value:
+            if v.HasField('simple_value'):
+                if event.step % _LOG_EVERY_N_STEPS == 0:
+                    _thread_pool.submit(_add_to_queue, key=v.tag,
+                                        value=v.simple_value, step=event.step,
+                                        run_id=mlflow.active_run().info.run_id)
+
+
+def _get_tensorboard_callback(lst):
+    for x in lst:
+        if isinstance(x, tensorflow.keras.callbacks.TensorBoard):
+            return x
+    return None
+
+
+def setup_callbacks(lst):
+    """
+    Adds TensorBoard and MlfLowTfKeras callbacks to the
+    input list, and returns the new list and appropriate log directory.
+    """
+    tb = _get_tensorboard_callback(lst)
+    if tb is None:
+        log_dir = 'tensorboard_logs'
+        l = lst + [TensorBoard(log_dir)]
+    else:
+        log_dir = tb.log_dir
+        l = lst
+    l += [__MLflowTfKerasCallback()]
+    return l, log_dir
+
+
+def autolog(metrics_every_n_steps=100):
```

Comment:
We should remember to add a docstring here (it'll show hosted API docs, e.g. [see the docs](https://mlflow.org/docs/latest/python_api/mlflow.tensorflow.html#mlflow.tensorflow.load_model) for `mlflow.tensorflow.load_model` and the [corresponding code](https://github.com/mlflow/mlflow/blob/0539a16f841f990717b3d1260e9670074e30a71c/mlflow/tensorflow.py#L181))